### DISCUSSION TITLE
Lint against unconditional recursion across function calls

### DISCUSSION THREAD

**Discussion by phil-opp** [rust-lang/rust-clippy#8446] [Ideas]
Rust's `unconditional_recursion` lint only works when a function calls itself _directly_. Thus it misses many cases of infinite recursions, including simple ones that can easily happen by accident. For example:

```rust
struct Foo;

impl From<u64> for Foo {
    fn from(v: u64) -> Self {
        v.into()
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=de26960a2c857050a67269b86ce9129f))

There are more examples in the tracking issue at https://github.com/rust-lang/rust/issues/57965. Ideally, this should be fixed in `rustc` itself of course, which was tried in https://github.com/rust-lang/rust/pull/75067. Unfortunately, the PR was closed because of its performance regressions.

Given that stack overflow errors are always fatal and can be hard to debug, I thought that such a lint might be a good candidate for clippy, where performance doesn't matter that much.

**Comment by llogiq** [comment]
I don't doubt that it's a worthwhile extension, however there are two things to keep in mind:

1. Cycles can form across arbitrary methods. The lint should only fire if the cycle is unconditional though.
2. We may not have full information: Some functions may be in external crates, perhaps even in generic code, in FFI code (and thus outside Rust) or dispatched dynamically (and thus not known at compile time).
